frontend/src/components/UserUpdateRoles.vue

```diff
@@ -1,210 +1,384 @@
+              </div>
+            </v-card>
+          </v-dialog>
+        </v-toolbar>
+      </template>
+      <template #item.lastLogin="{ item }">
+        <span v-if="item.lastLogin == -1">N/A</span>
+        <span v-else>
+          {{ new Date(item.lastLogin).toLocaleDateString("en-CA") }}
+        </span>
+      </template>
+
+      <template #item.roleArray="{ item }">
+        <span style="max-width:600px;display: flex;flex-wrap: wrap">
+          <span v-for="(val, index) of item.roleArray" v-bind:key="val.name">
+            {{ val }}
+            <span v-if="index != Object.keys(item.roleArray).length - 1"
+              >,
+            </span>
+          </span>
         </span>
-      </div>
-    </div>
+      </template>
 
+      <template #item.actions="{ item }">
+        <v-icon small @click="editRoles(item.clientRepresentation)">
+          mdi-pencil
+        </v-icon>
+      </template>
+    </v-data-table>
   </v-card>
 </template>
 
 <script>
 import ClientsRepository from "@/api/ClientsRepository";
 import UsersRepository from "@/api/UsersRepository";
-import UserUpdateSfdsAuth from "@/components/UserUpdateSfdsAuth";
 
 export default {
   name: "UserUpdateRoles",
-  components: { UserUpdateSfdsAuth },
-  props: ['userId'],
+  props: ["userId"],
   data() {
     return {
-      clients: [],
+      dialog: false,
+      headers: [
+        { text: "Application", value: "clientRepresentation.clientId" },
+        { text: "Role", value: "roleArray" },
+        { text: "Last Log In", value: "lastLogin"},
+        { text: "Actions", value: "actions", sortable: false },
+      ],
+      clientWithoutRoles: [],
       selectedClient: null,
       clientRoles: [],
       effectiveClientRoles: [],
-      selectedRoles: []
+      selectedRoles: [],
+      ClientsWithEffectiveRoles: [],
+      rolesLoaded: false,
+      isEdit: false,
     };
   },
   async created() {
-    //TODO error handling
-    await this.getClients();
+   this.loadUserRoles();
   },
   computed: {
     itemsInColumn() {
-      return Math.ceil(this.clientRoles.length / this.numberOfClientRoleColumns);
+      return Math.ceil(
+        this.clientRoles.length / this.numberOfClientRoleColumns
+      );
     },
     numberOfClientRoleColumns() {
-      return (this.clientRoles.length > 10) ? 2 : 1
+      return this.clientRoles.length > 10 ? 2 : 1;
     },
-    hasRoleForManageUserRoles: function() {
+    hasRoleForManageUserRoles: function () {
       const umsClientId = "USER-MANAGEMENT-SERVICE";
       const manageUserRolesName = "manage-user-roles";
 
-      return !!this.$keycloak.tokenParsed.resource_access[umsClientId].roles.includes(manageUserRolesName)
-    }
+      return !!this.$keycloak.tokenParsed.resource_access[
+        umsClientId
+      ].roles.includes(manageUserRolesName);
+    },
   },
   methods: {
-    roleArrayPosition: function(col, item) {
-      return (col - 1) * (this.itemsInColumn) + item - 1;
+    addRoles() {
```

Comment:
We've been using the `:function () { .. }` format in most other places. For consistency we should do the same here.